src/TriggerBinding/SqlTriggerListener.cs

```diff
@@ -0,0 +1,364 @@
+            {
+                this._changeMonitor.Stop();
+                this._state = State.Stopped;
+            }
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Returns the OBJECT_ID of userTable
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown if the query to retrieve the OBJECT_ID of the user table fails to correctly execute
+        /// This can happen if the OBJECT_ID call returns NULL, meaning that the user table might not exist in the database
+        /// </exception>
+        private async Task<int> GetUserTableIdAsync(SqlConnection connection, CancellationToken cancellationToken)
+        {
+            string getObjectIdQuery = $"SELECT OBJECT_ID(N{this._userTable.QuotedName}, 'U');";
+
+            using (var getObjectIdCommand = new SqlCommand(getObjectIdQuery, connection))
+            {
+                using (SqlDataReader reader = await getObjectIdCommand.ExecuteReaderAsync(cancellationToken))
+                {
+
+                    // TODO: Check if the below if-block ever gets hit.
+                    if (!await reader.ReadAsync(cancellationToken))
+                    {
+                        throw new InvalidOperationException($"Failed to determine the OBJECT_ID of the user table {this._userTable.FullName}");
+                    }
+
+                    object userTableId = reader.GetValue(0);
+
+                    if (userTableId is DBNull)
+                    {
+                        throw new InvalidOperationException($"Failed to determine the OBJECT_ID of the user table {this._userTable.FullName}. " +
+                            "Possibly the table does not exist in the database.");
+                    }
+
+                    return (int)userTableId;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the names and types of primary key columns of the user's table.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown if no primary keys are found for the user table. This could be because the user table does not have
+        /// any primary key columns.
+        /// </exception>
+        private async Task<IReadOnlyList<(string name, string type)>> GetPrimaryKeyColumnsAsync(SqlConnection connection, CancellationToken cancellationToken)
+        {
+            string getPrimaryKeyColumnsQuery = $@"
+                SELECT c.name, t.name, c.max_length, c.precision, c.scale
+                FROM sys.indexes AS i
+                INNER JOIN sys.index_columns AS ic ON i.object_id = ic.object_id AND i.index_id = ic.index_id
+                INNER JOIN sys.columns AS c ON ic.object_id = c.object_id AND ic.column_id = c.column_id
+                INNER JOIN sys.types AS t ON c.user_type_id = t.user_type_id
+                WHERE i.is_primary_key = 1 AND i.object_id = OBJECT_ID(N{this._userTable.QuotedName}, 'U');
```

Comment:
This needs to be `QuotedFullName`, otherwise you could get the wrong one if there's multiple tables with that name with different schemas.

Reply:
Although you could also probably simplify this by using the table ID retrieved in the call before this directly.

Reply:
Fixed. Now using table ID queried from here in the other two queries inside class `SqlTriggerListener`.